src/realm/query_expression.hpp

```diff
@@ -1989,75 +1990,57 @@ class Columns<BinaryData> : public SimpleQuerySupport<BinaryData> {
-
-template <>
-class Columns<StringData> : public SimpleQuerySupport<StringData> {
+class StringCompare {
 public:
-    using SimpleQuerySupport::SimpleQuerySupport;
-
-    Query equal(StringData sd, bool case_sensitive = true)
-    {
-        return string_compare<StringData, Equal, EqualIns>(*this, sd, case_sensitive);
-    }
-
-    Query equal(const Columns<StringData>& col, bool case_sensitive = true)
-    {
-        return string_compare<Equal, EqualIns>(*this, col, case_sensitive);
-    }
-
-    Query not_equal(StringData sd, bool case_sensitive = true)
-    {
-        return string_compare<StringData, NotEqual, NotEqualIns>(*this, sd, case_sensitive);
-    }
-
-    Query not_equal(const Columns<StringData>& col, bool case_sensitive = true)
+    explicit StringCompare(const Subexpr2<StringData>& source)
+        : m_source(source)
     {
-        return string_compare<NotEqual, NotEqualIns>(*this, col, case_sensitive);
-    }
 
-    Query begins_with(StringData sd, bool case_sensitive = true)
-    {
-        return string_compare<StringData, BeginsWith, BeginsWithIns>(*this, sd, case_sensitive);
     }
 
-    Query begins_with(const Columns<StringData>& col, bool case_sensitive = true)
-    {
-        return string_compare<BeginsWith, BeginsWithIns>(*this, col, case_sensitive);
-    }
+    StringCompare(const StringCompare&) = delete;
 
-    Query ends_with(StringData sd, bool case_sensitive = true)
-    {
-        return string_compare<StringData, EndsWith, EndsWithIns>(*this, sd, case_sensitive);
-    }
+    Query equal(StringData sd, bool case_sensitive = true);
+    Query equal(const Columns<StringData>& col, bool case_sensitive = true);
+    Query not_equal(StringData sd, bool case_sensitive = true);
+    Query not_equal(const Columns<StringData>& col, bool case_sensitive = true);
+    Query begins_with(StringData sd, bool case_sensitive = true);
+    Query begins_with(const Columns<StringData>& col, bool case_sensitive = true);
+    Query ends_with(StringData sd, bool case_sensitive = true);
+    Query ends_with(const Columns<StringData>& col, bool case_sensitive = true);
+    Query contains(StringData sd, bool case_sensitive = true);
+    Query contains(const Columns<StringData>& col, bool case_sensitive = true);
+    Query like(StringData sd, bool case_sensitive = true);
+    Query like(const Columns<StringData>& col, bool case_sensitive = true);
 
-    Query ends_with(const Columns<StringData>& col, bool case_sensitive = true)
-    {
-        return string_compare<EndsWith, EndsWithIns>(*this, col, case_sensitive);
-    }
-
-    Query contains(StringData sd, bool case_sensitive = true)
-    {
-        return string_compare<StringData, Contains, ContainsIns>(*this, sd, case_sensitive);
-    }
+private:
+    const Subexpr2<StringData>& m_source;
+};
 
-    Query contains(const Columns<StringData>& col, bool case_sensitive = true)
+template <>
+class Columns<StringData> : public SimpleQuerySupport<StringData>, public StringCompare {
```

Comment:
It would be nice to avoid the multiple inheritance design if possible. For one thing it leads to some unconventional lines like `StringCompare(*(static_cast<Subexpr2<StringData>*>(this)))` using `this` in the constructors. Is it possible to move all the functionality of the `StringCompare` class into a specialisation of `Subexpr2<StringData>` ? I think that would give you what you need both in `Columns<StringData>` and `ListColumns<StringData>`. 
(`Columns<StringData> --> SimpleQuerySupport<StringData> --> Subexpr2<StringData>`, and `ListColumns<StringData> --> ListColumnsBase<StringData> --> Subexpr2<StringData>`)